components/sup/src/manager/sup_watcher.rs

```diff
@@ -1,63 +1,70 @@
-                                             unwrap_or_default()).
-                                             unwrap_or_else(|_| PackageTarget::active_target());
-        if target == AARCH64_DARWIN {
-            Ok(SupWatcher::Fallback(PollWatcher::new_raw(tx).unwrap()))
-        } else {
-            Ok(SupWatcher::Native(RecommendedWatcher::new_raw(tx).unwrap()))
-        }
-    }
-
-    fn new(tx: Sender<DebouncedEvent>, delay: Duration) -> Result<Self> {
+    fn new<F: EventHandler>(event_handler: F, config: Config) -> Result<Self> {
         let target = PackageTarget::from_str(&env::var("HAB_STUDIO_HOST_ARCH").
                                              unwrap_or_default()).
                                              unwrap_or_else(|_| PackageTarget::active_target());
         if target == AARCH64_DARWIN {
             debug!("Using pollwatcher");
-            Ok(SupWatcher::Fallback(PollWatcher::new(tx, delay).unwrap()))
+            Ok(SupWatcher::Fallback(PollWatcher::new(event_handler, config).unwrap()))
         } else {
             debug!("Using native watcher");
-            Ok(SupWatcher::Native(RecommendedWatcher::new(tx, delay).unwrap()))
+            Ok(SupWatcher::Native(RecommendedWatcher::new(event_handler, config).unwrap()))
         }
     }
 
-    fn watch<P: AsRef<Path>>(&mut self, path: P, recursive_mode: RecursiveMode) -> Result<()> {
+    fn watch(&mut self, path: &Path, recursive_mode: RecursiveMode) -> Result<()> {
         match self {
             SupWatcher::Native(watcher) => watcher.watch(path, recursive_mode),
             SupWatcher::Fallback(watcher) => watcher.watch(path, recursive_mode),
         }
     }
 
-    fn unwatch<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
+    fn unwatch(&mut self, path: &Path) -> Result<()> {
         match self {
             SupWatcher::Native(watcher) => watcher.unwatch(path),
             SupWatcher::Fallback(watcher) => watcher.unwatch(path),
         }
     }
+
+    // For now we are using the default implementation of configure() provided
+    // by the notify crate which returns Ok(false) signalling that runtime
+    // configuration is not supported.
+
+    fn kind() -> WatcherKind
+        where Self: Sized
+    {
+        // https://github.com/notify-rs/notify/pull/441#discussion_r961970946
+        // Lacking a self reference I don't see how it isn't a mistake to include this method in the
+        // trait.  Trying to come up with an implementation I went and searched the notify-rs github
+        // repo an found the discussion linked above.  That's my best indicator that the maintainers
+        // agree with me.  So to satify this API I'm doing the following which is ugly and feels
+        // horrible so please do suggest or implement something better if you have something.
+        warn!("This implementation of kind() LIES to you. See comment in source code.");
+        WatcherKind::NullWatcher
```

Comment:
I think I would use an `unimplemented!` here. We are our only caller and should not be calling `kind`. This way we don't have to lie.

Reply:
Done but not yet pushed.